spatialdata/_core/_spatialdata.py

```diff
@@ -1,59 +1,83 @@
 from ome_zarr.io import parse_url
-from ome_zarr.reader import Reader
 
 from spatialdata._core.mixin.io_mixin import IoMixin
-from spatialdata._core.writer import write_spatial_anndata
-from spatialdata._core.transform import Transform, set_transform, get_transform
+
+# from spatialdata._core.writer import write_spatial_anndata
+from spatialdata._core.transform import Transform, get_transform, set_transform
+from spatialdata._core.writer import (
+    write_image,
+    write_labels,
+    write_points,
+    write_shapes,
+    write_tables,
+)
 from spatialdata.utils import are_directories_identical
 
 
 class SpatialData(IoMixin):
     """Spatial data structure."""
 
-    adata: AnnData
-    regions: Mapping[str, Any]
-    images: Mapping[str, Any]
-    points: Optional[AnnData]
+    tables: Optional[AnnData]
+    labels: Optional[Mapping[str, Any]]
+    images: Optional[Mapping[str, Any]]
+    points: Optional[Mapping[str, AnnData]]
+    shapes: Optional[Mapping[str, AnnData]]
+    elems: List[Optional[Set[str]]]
 
     def __init__(
         self,
-        adata: Optional[AnnData] = None,
-        regions: Mapping[str, Any] = MappingProxyType({}),
-        images: Mapping[str, Any] = MappingProxyType({}),
+        tables: Optional[AnnData] = None,
+        labels: Optional[Mapping[str, Any]] = MappingProxyType({}),
+        labels_transform: Optional[Mapping[str, Any]] = None,
+        images: Optional[Mapping[str, Any]] = MappingProxyType({}),
         images_transform: Optional[Mapping[str, Any]] = None,
-        points: Optional[AnnData] = None,
+        points: Optional[Mapping[str, AnnData]] = None,
+        points_transform: Optional[Mapping[str, Any]] = None,
+        shapes: Optional[Mapping[str, AnnData]] = None,
+        shapes_transform: Optional[Mapping[str, Any]] = None,
     ) -> None:
-        # current limitations:
-        # - only 2d data
-        # - only first table, image, regions of the dict are used
-        self.adata = adata
-        self.regions = dict(regions)
-        if images_transform is None:
-            images_transform = {k: Transform(ndim=2) for k in images}
-        assert set(images.keys()).issuperset(set(images_transform.keys()))
-        for k, v in images.items():
-            images_transform[k] = get_transform(v)
-        self.images = {
-            k: self.parse_image(image, image_transform)
-            for k, (image, image_transform) in zip(images.keys(), zip(images.values(), images_transform.values()))
-        }
-        self.points = points
+
+        elems = []
+
+        images, images_transform, elem_images = _validate_dataset(images, images_transform)
+        labels, labels_transform, elem_labels = _validate_dataset(labels, labels_transform)
+        points, points_transform, elem_points = _validate_dataset(points, points_transform)
+        shapes, shapes_transform, elem_shapes = _validate_dataset(shapes, shapes_transform)
+
+        if images is not None and images_transform is not None:
```

Comment:
it should never happen that one is None and the other not, I would replace this by 
if (... is None) + (... is None) % 2 == 0:
    code in the if block
else:
    exception
    
same for the ifs belows

Reply:
makes sense, updarted